docs/performance/productizing.rst

```diff
@@ -0,0 +1,116 @@
+.. _topics-to-watch-out-when-productizing-cratedb:
+
+###############################################
+ Topics to watch out when productizing CrateDB
+###############################################
+
+.. _mindful-of-memory:
+
+*******************************
+ Be mindful of memory capacity
+*******************************
+
+In CrateDB, operations requiring a working set like groupings, aggregations, and
+sorting are performed fully in memory without spilling over to disk.
+
+Sometimes you may have a query that leads to a sub-optimal execution plan
+requiring lots of memory. If you are coming to CrateDB from other database
+systems, your experience may be that these queries will proceed to run taking
+longer than required and impacting other workloads in the meanwhile. Sometimes
+this effect may be obvious if a query takes a lot of resources and runs for a
+long time, other times it may go unnoticed if a query that could complete in say
+100 milliseconds takes one hundred times longer, 10 seconds, but the users put
+up with it without reporting to you.
+
+In CrateDB if a query would require more heap memory than the interested nodes
+have available the query will fail with a particular type of error message that
+we call a `CircuitBreakerException`, this is a fail-fast approach, we can see
+quickly that there is an issue and optimize the query to get the best
+performance, without impacting other workloads.
+
+Please take a look at :ref:`Query Optimization 101 <performance-optimization>` for strategies to optimize your
+queries when you encounter this situation.
+
+.. _reading-lots-of-records:
+
+*************************
+ Reading lots of records
+*************************
+
+When the HTTP endpoint is used CrateDB will prepare the entire response in
+memory before sending it to the client.
+
+When the PostgreSQL protocol is used CrateDB attempts to stream the results but
+in many cases it still needs to bring all rows to the query handler node first.
```

Comment:
Why *attempts*? With PG, it will always stream data. Also inter-node data is paged, depending on the query (e.g the handler node does not have to apply any pipeline breaker like an final in-memory ordering), the data is also paged directly to the client. Thus, the handler node does not need to hold the complete result set in-memory in most cases.